Restore clip's cb() to its rightful glory - extract common debugging elements in llama

pwilkin opened the pull request:
I used my callback function from my Qwen3Next testing days, it seems like it works more cleanly than the previous one which was causing some problems with the scheduler / buffers.

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -2004,16 +2008,67 @@ struct clip_graph {
+        std::string t_name = std::string(name) + "_" + std::to_string(il);
+        ggml_tensor * args[] = { t };
+        ggml_tensor * res = ggml_custom_4d(ctx0, t->type, t->ne[0], t->ne[1], t->ne[2], t->ne[3], args, 1, print_debug, 1, nullptr);
+        strcpy(res->name, t_name.c_str());
```
use `ggml_set_name` instead

ngxson reviewed (commented):
If you want to go a step ahead, I would suggest using `ggml_backend_sched_set_eval_callback` to make it works the same way as libllama. This will be a cleaner solution

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -8,15 +8,19 @@
 #include <cstring>
 #include <fstream>
 #include <map>
+#include <memory>
```
some of these are already included by `clip-impl.h` - do we really need to include them again here?

pwilkin committed:
Extract common debugging functions; plug eval-callback and mtmd's MTMD_DEBUG_GRAPH with same functionality

pwilkin commented:
All right, based on the convo with @ngxson I've decided to tackle this properly:
* I moved the common debugging functions to `llama-debug.cpp`, added their headers to `llama.h` or `llama-cpp.h` depending on whether they use C or C++ APIs.
* I plugged `eval-callback` to those new common functions
* I modified mtmd's `cb()` to do the same thing as the `llm_graph_builder`'s one, which is basically to just set the tensor name. The entire tensor dump is set via `ggml_backend_sched_set_eval_callback`
* The added bonus is I created a template version of the `ggml_debug` function, so you can now set in the template whether NaNs should abort execution or not (default: no)

pwilkin committed:
Move to common

pwilkin committed:
Remove unneeded header

pwilkin committed:
Unlink from common

pwilkin committed:
chore: update webui build output

pwilkin committed:
Cleanup; properly pass params to mtmd without depending on common; factorize debug.cpp to use common debug code.

pwilkin committed:
Revert change to webapp

pwilkin committed:
Post-merge adjust

pwilkin committed:
Apply suggestions from code review

Co-authored-by: Xuan-Son Nguyen <thichthat@gmail.com>

pwilkin committed:
Apply code review changes

ngxson commented on tools/mtmd/clip.cpp:
```diff
@@ -1519,8 +1519,8 @@ struct clip_model_loader {
-                    model.mm_boi = get_tensor(string_format(TN_TOK_GLM_BOI, "weight"));
-                    model.mm_eoi = get_tensor(string_format(TN_TOK_GLM_EOI, "weight"));
+                    model.mm_boi = get_tensor(string_format(TN_TOK_GLM_BOI));
+                    model.mm_eoi = get_tensor(string_format(TN_TOK_GLM_EOI));
```
```suggestion
                    model.mm_boi = get_tensor(TN_TOK_GLM_BOI);
                    model.mm_eoi = get_tensor(TN_TOK_GLM_EOI);
```

ngxson commented on tools/server/server-context.cpp:
seems like any changes here can be reverted

pwilkin replied:
Yep.

pwilkin commented on tools/mtmd/clip.cpp:
```diff
@@ -3881,9 +3867,9 @@ void clip_debug_encode(clip_ctx * ctx, int h, int w, float fill_value) {
     }
-    bool cur_debug_graph = ctx->debug_graph;
-    ctx->debug_graph = true;
+    // TODO convert to new schema without breaking common separation?
```
The debug is broken here at the moment. And I can't access the definitions here without accessing `common`.

pwilkin committed:
Remove changes to server-context

pwilkin committed:
Remove mtmd.h include

pwilkin committed:
Remove utility functions from header

pwilkin commented on tools/mtmd/clip.cpp:
```diff
@@ -9,6 +9,7 @@
 #include "ggml-alloc.h"
 #include "ggml-backend.h"
 #include "gguf.h"
+#include "mtmd.h"
```
Done.

ngxson commented on common/debug.h:
```diff
@@ -0,0 +1,43 @@
+// ne   - the tensor dimensions array
+// nb   - the tensor strides array
+// n    - the number of rows/columns to fully print
+template <bool abort_on_nan> void common_ggml_print_tensor(uint8_t * data, ggml_type type, const int64_t * ne, const size_t * nb, int64_t n);
```
this function is not currently used outside `debug.cpp`, so probably it should also be private/static function

pwilkin replied:
Think this one might be useful on its own, so I think it might be worthwhile to keep it as a standalone utility function.

ngxson commented on common/debug.h:
```diff
@@ -0,0 +1,43 @@
+// The template parameter determins whether an error should be thrown whenever a NaN is encountered
+// in a tensor (useful for stopping debug sessions on first erroneous tensor)
+// The callback data will be passed as the third parameter (user_data)
+template <bool abort_on_nan> bool common_ggml_debug(struct ggml_tensor * t, bool ask, void * user_data);
```
the naming convention of libcommon is `common_(module)_(function)`, so I think this should be renamed to `common_debug_cb_eval`

pwilkin commented on tools/mtmd/clip.cpp:
```diff
@@ -3881,9 +3866,9 @@ void clip_debug_encode(clip_ctx * ctx, int h, int w, float fill_value) {
-    ctx->debug_graph = true;
+    // TODO convert to new schema without breaking common separation?
+    ctx->cb_eval = nullptr;
+    ctx->cb_eval_user_data = nullptr;
```
I mean your function manually set up some debugging callbacks, but now setting up debugging callbacks is done in the cli, so I'm wondering how to keep your debugging functionality.

pwilkin committed:
Apply suggestions from code review

Co-authored-by: Xuan-Son Nguyen <thichthat@gmail.com>

pwilkin committed:
Rename functions

pwilkin committed:
Update tools/mtmd/clip.cpp

Co-authored-by: Xuan-Son Nguyen <thichthat@gmail.com>

pwilkin committed:
Update tools/mtmd/clip.cpp

Co-authored-by: Xuan-Son Nguyen <thichthat@gmail.com>

pwilkin committed:
Update tools/mtmd/clip.cpp

Co-authored-by: Xuan-Son Nguyen <thichthat@gmail.com>

ngxson reviewed (approved):
Just tested it, looking good.

One small thing that I would prefer to have in this PR or a follow up one: the printed tensor has a long leading space, making it hard to read. Not sure why it's there in the first place, but better to remove it (not sure why it's there in the first place)

```
                                      [
                                       [      0.0000,      -0.0659,      -0.1201, ...,      -0.1427,      -0.1092,      -0.0488],
                                       [     -0.1201,      -0.1175,      -0.0603, ...,      -0.0488,      -0.0956,      -0.1014],
                                       [     -0.0603,      -0.1521,      -0.1467, ...,      -0.1014,      -0.1196,      -0.0856],
                                       ..., 
                                       [     -0.1674,      -0.0987,      -0.0784, ...,      -0.0787,      -0.1005,      -0.1428],
                                       [     -0.0784,      -0.1367,      -0.1161, ...,      -0.1428,      -0.1068,      -0.1274],
                                       [     -0.1161,      -0.0998,      -0.1306, ...,      -0.1274,       0.0000,       0.0000],
                                      ],
```
